File: src/com/gmail/marcosav2010/communicator/BaseCommunicator.java
package com.gmail.marcosav2010.communicator;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import com.gmail.marcosav2010.tasker.TaskOwner;

/**
 * This communicator sends and receives raw bytes, with a fixed buffer length.
 * 
 * @author Marcos
 *
 */
public class BaseCommunicator extends Communicator {

	public synchronized byte[] read(int bytes) throws IOException {
		byte[] read = getIn().readNBytes(bytes);

		return read;
	}

	public byte[] read(int bytes, TaskOwner taskOwner, long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
		return taskOwner.getExecutorService().submit(() -> read(bytes)).get(timeout, unit);
	}

	@Override
	public void write(byte[] bytes) throws IOException {
		getOut().write(bytes);
	}

	@Override
	public void close() throws IOException {
		if (getIn() != null)
			getIn().close();

		if (getOut() != null)
			getOut().close();
	}

	@Override
	public void closeQuietly() {
		try {
			close();
		} catch (IOException ex) {
		}
	}
}

File: core/src/main/java/com/gmail/marcosav2010/communicator/BaseCommunicator.kt
package com.gmail.marcosav2010.communicator

import com.gmail.marcosav2010.tasker.TaskOwner
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * This communicator sends and receives raw bytes, with a fixed buffer length.
 *
 * @author Marcos
 */
open class BaseCommunicator : Communicator() {

    @Synchronized
    open fun read(bytes: Int): ByteArray? = input!!.readNBytes(bytes)

    open fun read(bytes: Int, taskOwner: TaskOwner, timeout: Long, unit: TimeUnit): ByteArray? =
            taskOwner.executorService.submit<ByteArray?> { read(bytes) }[timeout, unit]

    override fun write(bytes: ByteArray) = output!!.write(bytes)

    override fun close() {
        input?.close()
        output?.close()
    }

    override fun closeQuietly() {
        try {
            close()
        } catch (ignored: IOException) {
        }
    }
}